tech/post/oslo_privsep.md

```diff
@@ -0,0 +1,176 @@
+```
+ploop: CommandFilter, ploop, root
+[name]: CommandFilter, [command], [user]
+```
+2. **RegExpFilter**: Generic filter that checks the executable called, then uses a list of regular expressions to check all subsequent arguments
+```
+tunctl: RegExpFilter, /usr/sbin/tunctl, root, tunctl, -b, -t, .*
+[name]: RegExpFilter, [command], [user], [command_regex, [option],...]
+```
+
+3. **PathFilter**: Generic filter that lets you check that paths provided as parameters fall under a given directory
+```
+chown: PathFilter, /bin/chown, root, nova, /var/lib/images
+[name]: PathFilter, [command], [user], [path]
+```
+
+4. **EnvFilter**: Filter allowing extra environment variables to be set by the calling code
+```
+dnsmasq: EnvFilter, env, root, CONFIG_FILE=, NETWORK_ID=, dnsmasq
+[name]: EnvFilter, env, [user], [env_config1, [env_config2]...] [command]
+```
+It also have ``ReadFileFilter``, ``IPFilter`` and ``IpNetnsExecFilter`` commands.
+So back to our specific case, we need to add one more filter if we
+want to execute the previous command.
+````
+[Filters]
+# cinder/volume/iscsi.py: iscsi_helper '--op' ...
+tgtadm: CommandFilter, tgtadm, root
+````
+When ``rootwrap`` is been executed, the program will parse the config file, load the filters and valid the input command at the very begining,and then create a sub-process to execute the command.
+
+Based on the process above, we could find the disadvantage of rootwrap:
+
+1. Efficiency: Generating command lines and parsing textual output from tools is slow.
+2. Convenience: Service could have lots of privileged commands and it's hard to configure those filters correctly, because the filters have limited expressiveness.
+3. Security: ``chown: CommandFilter, chown, root`` this command allows
+the invoking user to run ``chown`` with any arguments as root. it could be dangerous if ``/etc/shadow`` is been passed.
+
+Rootwrap team has tried to improve the efficiency by adding the
+[daemon](https://specs.openstack.org/openstack/oslo-specs/specs/liberty/privsep.html) support in Liberty, but it still can not fix the Convenience and Security issues.
+
+## Oslo.privsep
+
+Privsep is designed based on the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege). It provides a fine-grained security control by Linux Capabilities rather than run the process as a superuser directly, and the client communicate with the privileged daemon through ``unix domain socket``, which is better than command line.
+
+### Linux capabilities
+
+The feature is added since kernel 2.2, linux divides the privileges traditionally associated with superuser into distinct units, known as capabilities, which can be independently enabled and disabled.  It's a per-thread attribute. 
+Each thread has three capability sets containing zero or more of the capabilities.
+1. **Permitted**: This is a limiting superset for the effective capabilities that the thread may assume.
```

Comment:
S/a limiting superset for/a limiting superset of

Reply:
done